lib/utils.js

```diff
@@ -30,4 +31,116 @@ function inspectObject (args) {
   }
 }
 
-export { inspectObject };
+/**
+ * Takes the caps that were provided in the request and translates them
+ * into caps that can be used by the inner drivers.
+ * @param {*} jsonwpCaps
+ * @param {*} w3cCapabilities
+ * @param {*} constraints
+ * @param {*} defaultCapabilities
+ */
+function parseCapsForInnerDriver (jsonwpCaps, w3cCapabilities, constraints={}, defaultCapabilities={}) {
+  // Check if the caller sent JSONWP caps, W3C caps, or both
+  const hasW3CCaps = _.isObject(w3cCapabilities);
+  const hasJSONWPCaps = _.isObject(jsonwpCaps);
+
+  // Make copies of the capabilities that include the default capabilities
+  if (hasW3CCaps) {
+    w3cCapabilities = {
+      ...w3cCapabilities,
+      alwaysMatch: {
+        ...defaultCapabilities,
+        ...w3cCapabilities.alwaysMatch,
+      },
+    };
+  }
+
+  if (hasJSONWPCaps) {
+    jsonwpCaps = {
+      ...defaultCapabilities,
+      ...jsonwpCaps,
+    };
+  }
+
+  // Get MJSONWP caps
+  let desiredCaps = {};
+  if (hasJSONWPCaps) {
+    desiredCaps = jsonwpCaps;
+  }
+
+  // Get W3C caps
+  if (hasW3CCaps) {
+    // Call the process capabilities algorithm to find matching caps on the W3C (https://github.com/jlipps/simple-wd-spec#processing-capabilities)
+    let matchingW3CCaps = processCapabilities(w3cCapabilities, constraints, true);
+
+    if (hasJSONWPCaps) {
+      let jsonwpKeys = _.keys(desiredCaps);
+      let matchingW3CCapsKeys = _.keys(matchingW3CCaps);
+
+      // If JSONWP has keys that W3C doesn't have, log a warning message
+      let differingKeys = _.differenceWith(jsonwpKeys, matchingW3CCapsKeys);
+      if (differingKeys.length > 0) {
+        logger.warn(`The following capabilities were provided in the JSONWP desired capabilities but not the W3C capabilities: ${JSON.stringify(differingKeys)}`);
+      }
+    }
+
+    // If JSONWP caps and W3C caps were provided and there are discrepancies between them
+    // we'll be forgiving for now and just merge the two. But in the future, the clients
+    // should not be sending discrepant caps and Appium should ignore the extraneous MJSONWP caps
+    desiredCaps = {
+      ...desiredCaps, // TODO: This line should be removed once we decide to not allow discrepent caps
+      ...matchingW3CCaps,
+    };
+  }
+
+  let newJsonwpCaps, newW3CCapabilities;
+
+  // Create new JSONWP desired capabilities and W3C 'capabilities' objects that have the reconciled attributes
+  if (hasJSONWPCaps) {
+    newJsonwpCaps = {...desiredCaps};
+  }
+
+  // Translate the matched W3C caps back into W3C capabilities object so that it
+  // can be proxied to the inner driver
+  if (hasW3CCaps) {
+    newW3CCapabilities = {
+      alwaysMatch: {...insertAppiumPrefixes(desiredCaps)}, // Insert 'appium:' prefix to non-prefixed caps so that the Inner Driver doesn't complain about bad W3C prefixes
+      firstMatch: [{}],
+    };
+  }
+
+  return {desiredCaps, newJsonwpCaps, newW3CCapabilities};
```

Comment:
so it's ok if some of these are undefined?

Reply:
It is yeah. That's how drivers distinguish between JSONWP vs w3c

Reply:
I'd rather pass null instead of undefined